xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-templates/src/test/java/org/xwiki/web/DocumentTagsTest.java

```diff
@@ -103,31 +104,31 @@ void displayTagsWhenEditRightsAndTagPluginAvailableAndNoTags() throws Exception
         // Remove extra spaces to make it easy to assert the result below.
         String result = templateManager.render("documentTags.vm").trim().replaceAll("\\s+", " ");
 
-        // Verify that the generated HTML matches the expectation:
+        // Verify that the generated HTML matches the expectations:
         // - The tag label is displayed
         // - No tag is listed after the tag label
         // - The "+" link is displayed since the user has edit rights
         assertThat(result, matchesPattern("\\Q<div class=\"doc-tags\" id=\"xdocTags\"> core.tags.list.label "
-            + "<div class=\"tag-tool tag-add\"><a href=\"\\E.*\"\\Q title=\"core.tags.add.tooltip\" "
-            + "rel=\"nofollow\">[+]</a></div> </div>\\E"));
+            + "<div class=\"tag-tool tag-add\"> <a href=\"/xwiki/bin/view/space/page?showTagAddForm=true#xdocTags\" title=\"core.tags.add.tooltip\" "
+            + "rel=\"nofollow\">[+]</a> </div> </div>"));
```

Comment:
This shows that you've introduced some spaces. I would prefer if we could avoid that as I think spaces between the elements could be visible (haven't checked, though). Maybe you could check the Velocity and introduce some comments or put the line breaks at other places to avoid them?

Also, now that you've put the full URL in the string, I think you could/should use `assertEquals` instead of a regular expression.

Reply:
I checked and found out that these extra 2 spaces (they weren't there before), was the result of splitting the ```#if($hasedit && $xwiki.tag)``` even though I have put ```##``` at the end of HTML tags as suggested by @manuelleduc, I am not sure why the 2 extra space is being generated in the velocity.
The velocity needs to be changed now. Doing it...

Reply:
@wthrajat did you try by adding some more `##` on other newly introduced linebreaks? The nice thing with the page tests is that you can quickly iterate by editing the velocity and running the tests.